keps/sig-storage/1698-generic-inline-volumes/README.md

```diff
@@ -0,0 +1,540 @@
+
+#### Local LVM storage as scratch space
+
+Applications working with data sets that exceed the RAM size can
+request local storage with performance characteristics or size that is
+not met by the normal Kubernetes `EmptyDir` volumes. For example,
+[TopoLVM](https://github.com/cybozu-go/topolvm) was written for that
+purpose.
+
+### Risks and Mitigations
+
+Enabling this feature allows users to create PVCs indirectly if they can
+create pods, even if they do not have permission to create them
+directly. Cluster administrators must be made aware of this. If this
+does not fit their security model, they can disable the feature
+through the feature gate that will be added for the feature.
+
+Alternatively, a label on a namespace could be used to disable the
+feature just for that namespace.
+
+## Design Details
+
+### PVC meta data
+
+The namespace is the same as for the pod and the name is the
+concatenation of pod name and the `Volume.Name`. This is guaranteed to
+be unique for the pod and each volume in that pod. Care must be taken
+by the user to not exceed the length limit for object names, otherwise
+volumes cannot be created.
+
+Other meta data is copied from the `VolumeClaimTemplate.ObjectMeta`.
+
+### Preventing accidental collision with existing PVCs
+
+The controller will only create missing PVCs. It does not need to
+delete (handled by garbage collection) nor update (`PodSpec.Volumes`
+is immutable) existing PVCs. Therefore there is no risk that the
+controller will accidentally modify unrelated PVCs.
+
+The volume scheduling library and kubelet must check that a PVC is
+owned by the pod before using it. Otherwise they must ignored it. This
+ensures that a volume is not used accidentally for a pod in case of a
+name conflict.
+
+To surface such a name conflict to the user, the controller can also
+check for PVCs that aren't owned by the pod in its reconciliation loop
+and emit a pod event that describes the problem.
+
+### Feature gate
+
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
+- kubelet and scheduler accept such volumes.
+
+Existing pods with such a volumes will not be scheduled respectively
+started when the feature gate is off.
+
+### Modifying volumes
+
+Once the PVC for an inline volume has been created, it can be updated
+directly like other PVCs. The controller will not interfere with that
+because it never updates PVCs. This can be used to control features
+like volume resizing.
+
+### Late binding
+
+Ideally, provisioning should use late binding (aka
+`WaitForFirstConsumer`). The initial implementation assumes that the
+user is taking care of that by selecting that provisioning mode in the
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
```

Comment:
I couldn't find a reason why late binding should *not* be used. The volume is clearly going to be created for use by this one pod, so using late binding seems like the right thing to do (don't create volumes unless the pod can start, better scheduling) in all cases.

By enabling late binding for inline volumes, we reduce the amount of storage classes which have to be defined in the cluster. It might not be relevant, though, if all classes already have late binding enabled.